stage1/init/cgroup.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//+build linux
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/schema/types"
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/coreos/go-systemd/unit"
+	"github.com/coreos/rkt/common"
+)
+
+type addIsolatorFunc func(opts []*unit.UnitOption, limit string) ([]*unit.UnitOption, error)
+
+var (
+	isolatorFuncs = map[string]addIsolatorFunc{
+		"cpu":    addCpuLimit,
+		"memory": addMemoryLimit,
+	}
+	cgroupControllerRWFiles = map[string][]string{
+		"memory": []string{"memory.limit_in_bytes"},
+		"cpu":    []string{"cpu.cfs_quota_us"},
+	}
+)
+
+func addCpuLimit(opts []*unit.UnitOption, limit string) ([]*unit.UnitOption, error) {
+	milliCores, err := strconv.Atoi(limit)
+	if err != nil {
+		return nil, err
+	}
+	quota := strconv.Itoa(milliCores/10) + "%"
+	opts = append(opts, newUnitOption("Service", "CPUQuota", quota))
+	return opts, nil
+}
+
+func addMemoryLimit(opts []*unit.UnitOption, limit string) ([]*unit.UnitOption, error) {
+	opts = append(opts, newUnitOption("Service", "MemoryLimit", limit))
+	return opts, nil
+}
+
+func maybeAddIsolator(opts []*unit.UnitOption, isolator string, limit string) ([]*unit.UnitOption, error) {
+	var err error
+	if isIsolatorSupported(isolator) {
+		opts, err = isolatorFuncs[isolator](opts, limit)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		fmt.Fprintf(os.Stderr, "warning: resource/%s isolator set but support disabled in the kernel, skipping\n", isolator)
+	}
+	return opts, nil
+}
+
+func isIsolatorSupported(isolator string) bool {
+	if files, ok := cgroupControllerRWFiles[isolator]; ok {
+		for _, f := range files {
+			isolatorPath := filepath.Join("/sys/fs/cgroup/", isolator, f)
+			if _, err := os.Stat(isolatorPath); os.IsNotExist(err) {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+func parseCgroups() (map[int][]string, error) {
```

Comment:
I'd like to see this function take an io.Reader and then have some tests for it (then pass /proc/cgroups on 165)